enterprise_gateway/services/processproxies/k8s.py

```diff
@@ -197,15 +197,19 @@ def _create_kernel_namespace(self, service_account_name):
             # creating a role each time.
             self._create_role_binding(namespace, service_account_name)
         except Exception as err:
-            if self.delete_kernel_namespace:
-                reason = "Error occurred creating role binding for namespace '{}': {}".format(namespace, err)
-                # delete the namespace since we'll be using the EG namespace...
-                body = client.V1DeleteOptions(grace_period_seconds=0, propagation_policy='Background')
-                client.CoreV1Api().delete_namespace(name=namespace, body=body)
-                self.log.warning("Deleted kernel namespace: {}".format(namespace))
+            if isinstance(err, client.rest.ApiException) and err.status == 409 and self.kernel_manager.restarting:
+                self.delete_kernel_namespace = True  # okay if ns already exists and restarting, still mark for delete
```

Comment:
What is the case of this line?
As your fix lands, there's no namespace creation while restarting kernel, is there?

Reply:
@esevan - correct, we will not delete a kernel pod's namespace during the shutdown phase of a kernel restart so we will tolerate its existence (409) during the start phase of a kernel restart.

You may have noticed that Issue #578 references a `309` error code, not `409`.  Actually, I found both to occur during my analysis - due to timing.  `409` occurs if the object being created already exists, while `309` occurs if the object being created happens to be _terminating_.  I believe there was a race condition occurring relative to the namespace deletion that used to occur during the shutdown phase.  Since that deletion is no longer attempted, then `409` (already exists) is the error code we should tolerate if the kernel is restarting.

So the net result is that a kernel pod's namespace behaves identical to if the user brought their own namespace via `KERNEL_NAMESPACE` relative to kernel restarts.  However, we still want to _ultimately_ delete the namespace when the kernel is finally shutdown - so we set the instance's member `self.delete_kernel_namespace = True` - since we're in the process of creating a new instance of this process proxy.

Reply:
Okay I thought it was better to check if kernel is restarting before creating namespace, but I agree with timing issue can happen.

+1, Is there any contribution rule to approve PR?

Reply:
There's no rule for approval.  Your review is appreciated!

Reply:
@esevan just do a comment with LGTM, or even better, review and approve the PR via the Github UI.